Source/Core/VideoCommon/VertexLoaderARM64.cpp

```diff
@@ -209,12 +209,26 @@ int VertexLoaderARM64::ReadVertex(VertexComponentFormat attribute, ComponentForm
   // Z-Freeze
   if (native_format == &m_native_vtx_decl.position)
   {
-    CMP(count_reg, 3);
-    FixupBranch dont_store = B(CC_GT);
-    MOVP2R(EncodeRegTo64(scratch2_reg), VertexLoaderManager::position_cache);
-    ADD(EncodeRegTo64(scratch1_reg), EncodeRegTo64(scratch2_reg), EncodeRegTo64(count_reg),
-        ArithOption(EncodeRegTo64(count_reg), ShiftType::LSL, 4));
-    m_float_emit.STUR(write_size, coords, EncodeRegTo64(scratch1_reg), -16);
+    CMP(remaining_reg, 3);
+    FixupBranch dont_store = B(CC_GE);
+    MOVP2R(EncodeRegTo64(scratch2_reg), VertexLoaderManager::position_cache.data());
+    ADD(EncodeRegTo64(scratch1_reg), EncodeRegTo64(scratch2_reg), EncodeRegTo64(remaining_reg),
+        ArithOption(EncodeRegTo64(remaining_reg), ShiftType::LSL, 4));
+    m_float_emit.STUR(write_size, coords, EncodeRegTo64(scratch1_reg), 0);
```

Comment:
As far as I can tell, x64 is always writing 128 bits here, but AArch64 can write different sizes. Is there a particular reason for this? (Is it just for the sake of using less memory bandwidth?) If always writing 128 bits is fine, the optimization I suggested should work here as well (after replacing `STUR` with `STR`). Otherwise I suppose you could apply the optimization only when `write_size == 128`.

Reply:
See this:

https://github.com/dolphin-emu/dolphin/blob/3fdc6cbffea5f78dcfd8c2bad3cec12b3d45fdb1/Source/Core/VideoCommon/VertexLoaderManager.h#L54-L57

The values are padded to allow writing 4 values, even though only 3 of them are valid (I guess the 4th one is garbage?), and for XY-only vertices only 2 of them might be valid (the zfreeze code [zeros the third element in that case](https://github.com/dolphin-emu/dolphin/blob/3fdc6cbffea5f78dcfd8c2bad3cec12b3d45fdb1/Source/Core/VideoCommon/VertexManagerBase.cpp#L563-L564)).

For normals, I believe `count_out` will always be 3, an thus `write_size` will be 128.

I copied the code that already existed for positions, so I don't know exactly why it's implemented the way it is.

Reply:
I think I've implemented this, not 100% sure I did it right though.

(Also, do we need to add some kind of alignment attribute to `position_cache` to ensure it's 16-byte aligned?)

Reply:
OK, I've tried to implement this; however it's failing on the OSX buildbot for ARM:

> ```
>       Start 27: VertexLoaderTest
> 27/27 Test #27: VertexLoaderTest .................Bus error***Exception:   0.40 sec
> ```

Using `alignas(sizeof(std::array<float, 4>))` didn't help (and on further thought, I probably would get an alignment exception and not a bus error in that case).

-----

... and it turns out that the cause was actually that I was still using `scratch1_reg` for the `STR` when I had removed the `ADD` instruction; switching to `scratch2_reg` fixed it.  I still have some uncertainty as to whether the `alignas` is needed or not; I've left it in for now since it doesn't seem harmful.

Reply:
If it isn't 16-byte aligned, the code will still work but the performance may be lower depending on the CPU model. So having it properly aligned sounds like a good idea.